trieste/space.py

```diff
@@ -171,3 +172,20 @@ def discretize(self, num_samples: int) -> DiscreteSearchSpace:
             this :class:`Box`.
         """
         return DiscreteSearchSpace(points=self.sample(num_samples))
+
+    def __mul__(self, other_box: Box):
+        """
+        Combines 2 Box spaces into a single one by concatenating their respective lower and upper bounds.
+        :param other_box: :class:`Box`.
+        :return: the new combined :class:`Box`.
+        :raise ValueError (or InvalidArgumentError): If ``other_box`` is not from :class:`Box`.
+        """
+        if self.lower.dtype is not other_box.lower.dtype:
+            raise TypeError(
+                f"Bounds of both boxes must have the same dtype, got {self.lower.dtype} and"
+                f" {other_box.lower.dtype}"
+            )
```

Comment:
this is nice, but in this case we need to `return NotImplemented` for invalid args (see [the docs](https://docs.python.org/3/reference/datamodel.html#object.__mul__))

Reply:
you'll need to update the docstring accordingly

Reply:
is this what you had in mind?

Reply:
hehe, almost, but this has a funny bug. `NotImplemented` doesn't take arguments (it's a builtin constant not an exception type). So when you try to pass it arguments, python raises a `TypeError`, and your test passes